libp2p/services/autonatservice.nim

```diff
@@ -0,0 +1,128 @@
+# Nim-LibP2P
+# Copyright (c) 2022 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import std/[options, tables]
+import chronos
+import ../switch
+import ../protocols/[connectivity/autonat,
+                    rendezvous]
+import ../protocols/connectivity/relay/[relay, client]
+import ../discovery/[rendezvousinterface, discoverymngr]
+import ../utils/heartbeat
+import ../crypto/crypto
+
+type
+  AutonatService* = ref object of Service
+    registerLoop: Future[void]
+    scheduleInterval: Option[Duration]
+    networkReachability: NetworkReachability
+    t: CountTable[NetworkReachability]
+    autonat: Autonat
+    newStatusHandler: NewStatusHandler
+    rng: ref HmacDrbgContext
+    numPeersToAsk: int
+    maxConfidence: int
+
+
+  NetworkReachability* {.pure.} = enum
+    NotReachable, Reachable, Unknown
+
+  NewStatusHandler* = proc (networkReachability: NetworkReachability): Future[void]  {.gcsafe, raises: [Defect].}
+
+proc new*(
+  T: typedesc[AutonatService],
+  autonat: Autonat,
+  rng: ref HmacDrbgContext,
+  scheduleInterval: Option[Duration] = none(Duration),
+  numPeersToAsk: int = 5,
+  maxConfidence: int = 3): T =
+  return T(
+    scheduleInterval: scheduleInterval,
+    networkReachability: NetworkReachability.Unknown,
+    t: initCountTable[NetworkReachability](),
+    autonat: autonat,
+    rng: rng,
+    numPeersToAsk: numPeersToAsk,
+    maxConfidence: maxConfidence)
+
+proc networkReachability*(self: AutonatService): NetworkReachability {.inline.} =
+  return self.networkReachability
+
+proc handleAnswer(self: AutonatService, ans: NetworkReachability) {.async.} =
+  if ans == NetworkReachability.Unknown:
+    return
+  if ans == self.networkReachability:
+    if self.t[ans] == self.maxConfidence:
+      return
+    self.t.inc(ans)
+  else:
+    if self.t[self.networkReachability] > 0:
+      self.t.inc(self.networkReachability, -1)
+    if self.t[ans] < self.maxConfidence:
+      self.t.inc(ans)
+    if self.t[ans] == self.maxConfidence or self.t[self.networkReachability] == 0:
+      self.networkReachability = ans
+
+  if self.t[self.networkReachability] == self.maxConfidence:
```

Comment:
just a question, since im not sure I get this. When are we calling the `newStatusHandler`. I am expecting it to be called every time the status changes, so unsure how the `maxConfidence` applies here.

Reply:
The status can also change with zero confidence, I thought it wouldn't make sense to call in this case. Wdyt?

Reply:
mm I mean what i expected is that for every `interval` we call `newStatusHandler` with the new status. It can be `unknown`, `unreachable` or `rechable` depending on the result (which depends on `maxConfidence`) but is called no matter what.